onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -637,6 +644,56 @@ class PlannerImpl {
+              // get at this stage (TODO: Investigate feasibility of this, see TODO in FinalizeSessionStateImpl() around this)
+
+              // Instead, we take a "less than ideal" route which is to set the location to be the device
+              // corresponding to the EP that the node is partitioned to. The hypothesis is that it is "most likely"
+              // that the implicit input will eventually be consumed on that device in a nested subgraph.
+
+              // The previous behavior was to default to CPU which will cause unnecessary copies when
+              // (1) The user invokes Run() with an OrtValue backed by non-CPU memory (eg CUDA) and
+              // the explicit consumer of the implicit input is on the non-CPU device in the subgraph
+              // (2) The user tries to IOBind implicitly consumed graph inputs (GH Issue 11254) and
+              // the explicit consumer of the implicit input is on the non-CPU device in the subgraph
+
+              // Even if the user provides an input on CPU and the explicit consumer of the implicit input
+              // is on the non-CPU device, instead of the subgraph copying mechanism taking it to the device,
+              // all we will do is "front-load" this copy in utils::CopyInputsAcrossDevices() with this approach.
+
+              // NOTE 1: The only case this will be sub-optimal is when a node containing a subgraph is partitioned to a
+              // non-CPU EP and the user provides an input (or tries to IOBind the input) AND it will eventually be
+              // explicitly consumed on CPU - this scenario should be very rare and we forgo performance in this case
+              // (the subgraph copying mechanism will make the copy to CPU eventually) in favor of optimizing for the
+              // common case (which is that we expect the implicit input to be consumed on the non-CPU device corresponding
+              // to the non-CPU EP).
+
+              // NOTE 2: If the implicit input is consumed by multiple nodes (as implicit inputs in all of them) and
+              // all of them are partitioned to the same EP, then we go ahead with the above stated logic.
+              // If there are multiple EPs involved, we default the location to just CPU as there is ambiguity involved
+              // as to which non-CPU device is "most optimal" for the implicit input.
+
+              if (set_implicitly_consumed_node_arg_has_heterogenous_ep_consumers.count(index) == 0) {
+                // First time we are encountering this implicitly consumed input at this graph level (or)
+                // the EP that we previously seen for this implicit input is the same one as the current EP
+                // we have seen
+                if (!map_implicitly_consumed_node_arg_to_ep.contains(index)) {
+                  plan_.SetLocation(static_cast<size_t>(index), exec_provider->GetAllocator(0, OrtMemType::OrtMemTypeDefault)->Info());
```

Comment:
nit: could call 'find' once instead of calling 'contains' in the 'if' and 'find' in the 'else if'

Reply:
Done